source/sundials/kinsol.cc

```diff
@@ -147,8 +146,88 @@ namespace SUNDIALS
 
       return err;
     }
+
+#  else // SUNDIALS 5.0 or later
+
+    template <typename VectorType>
+    int
+    t_kinsol_setup_jacobian(N_Vector u,
+                            N_Vector f,
+                            SUNMatrix /* ignored */,
+                            void *user_data,
+                            N_Vector /* tmp1 */,
+                            N_Vector /* tmp2 */)
+    {
+      // Receive the object that describes the linear solver and
+      // unpack the pointer to the KINSOL object from which we can then
+      // get the 'setup' function.
+      const KINSOL<VectorType> &solver =
+        *static_cast<const KINSOL<VectorType> *>(user_data);
+
+      // Allocate temporary (deal.II-type) vectors into which to copy the
+      // N_vectors
+      GrowingVectorMemory<VectorType>            mem;
+      typename VectorMemory<VectorType>::Pointer ycur(mem);
+      typename VectorMemory<VectorType>::Pointer fcur(mem);
+      solver.reinit_vector(*ycur);
+      solver.reinit_vector(*fcur);
+
+      copy(*ycur, u);
+      copy(*fcur, f);
```

Comment:
This whole copy business can be replaced with the deal.II NVectorView (#11395 ) (in a future PR).

Reply:
Ah, I was unaware we had that. I read the paper that describes all of the new SUNDIALS stuff and had the impression that one can generate `N_Vector` objects that are just wrappers around everyone else's favorite vector type. But we don't use that here, and I thought we didn't have those implemented yet. I guess I have to read through #11395, unless you wanted to make the conversion yourself. In any case, like you say, this is a future problem.